include/swift/ABI/TaskGroup.h

```diff
@@ -0,0 +1,393 @@
+        }
+
+        /// Status value decrementing the Ready, Pending and Waiting counters by one.
+        GroupStatus completingPendingReadyWaiting() {
+          assert(pendingTasks() && "can only complete waiting task when pending tasks available");
+          assert(readyTasks() && "can only complete waiting task when ready tasks available");
+          assert(hasWaitingTask() && "can only complete waiting task when waiting task available");
+          return GroupStatus{status - waiting - oneReadyTask - onePendingTask};
+        }
+        GroupStatus completingPendingReady() {
+          assert(pendingTasks() && "can only complete waiting task when pending tasks available");
+          assert(readyTasks() && "can only complete waiting task when ready tasks available");
+          return GroupStatus{status - oneReadyTask - onePendingTask};
+        }
+
+        /// Pretty prints the status, as follows:
+        /// GroupStatus{ P:{pending tasks} W:{waiting tasks} {binary repr} }
+        std::string to_string() {
+          std::string str;
+          str.append("GroupStatus{ ");
+          str.append("C:"); // cancelled
+          str.append(isCancelled() ? "y " : "n ");
+          str.append("W:"); // has waiting task
+          str.append(hasWaitingTask() ? "y " : "n ");
+          str.append("R:"); // ready
+          str.append(std::to_string(readyTasks()));
+          str.append(" P:"); // pending
+          str.append(std::to_string(pendingTasks()));
+          str.append(" " + std::bitset<64>(status).to_string());
+          str.append(" }");
+          return str;
+        }
+
+        /// Initially there are no waiting and no pending tasks.
+        static const GroupStatus initial() {
+          return GroupStatus{0};
+        };
+    };
+
+    template<typename T>
+    class NaiveQueue {
+        std::queue<T> queue;
+
+    public:
+        NaiveQueue() = default;
+        NaiveQueue(const NaiveQueue<T> &) = delete ;
+        NaiveQueue& operator=(const NaiveQueue<T> &) = delete ;
+
+        NaiveQueue(NaiveQueue<T>&& other) {
+          queue = std::move(other.queue);
+        }
+
+        virtual ~NaiveQueue() { }
+
+        bool dequeue(T &output) {
+          if (queue.empty()) {
+            return false;
+          }
+          output = queue.front();
+          queue.pop();
+          return true;
+        }
+
+        void enqueue(const T item) {
+          queue.push(item);
+        }
+    };
+
+  private:
+
+//    // TODO: move to lockless via the status atomic
+    mutable std::mutex mutex;
+
+    /// Used for queue management, counting number of waiting and ready tasks
+    std::atomic<uint64_t> status;
+
+    /// TaskStatusRecord that is attached to the task running the group.
+    ///
+    /// Because we must remove it from the task as we exit/destroy the group,
+    /// we have to keep this pointer here so we know which record to remove then.
+    TaskGroupTaskStatusRecord* Record;
```

Comment:
Ah interesting... this had not occurred to me. I really only fully grasped how the records are to be used quite late in the development of groups, so perhaps this is a historical thing rather than a necessary thing to be honest.

I think you're right that it could be one and the same thing 🤔 It simplifies this weird thing that the group stores the record, it it simply would _be_ the record.

I can give that a shot.